src/ast/expr_ext.rs

```diff
@@ -0,0 +1,42 @@
+use crate::{
+    ast::{self, support::*, AstNode, AstToken},
+    kinds::SyntaxKind::*,
+    match_ast, SyntaxElement,
+};
+
+use super::support::first;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum LiteralKind {
```

Comment:
Could we simply define `enum Literal` like `enum Expr` and remove `NODE_LITERAL`?

Reply:
There is now a comment that explains

Reply:
:thinking: I don't consider `Str` is a *Literal* in my own parser crate. Because the interpolation making it not *literally* mean the text of it.
Well, that's arguable.

Reply:
What if we had a dedicated node for each of these which is otherwise just a token? e.g., NODE_INTEGER which just has a TOKEN_INTEGER child. Then we don't need to define these manually, and we don't need to answer the question of which of them are "literals".

Reply:
What if we get rid of the concept of a `Literal` and inline it into `Expr`. This would be the most simple. I don't feel like the literal concept is very useful seeing as different things can be interpreted as a literal. I think this is similar to how `Value` was not very useful.

Reply:
> What if we get rid of the concept of a Literal and inline it into Expr.

I'd stick on current node structure and not touching the parser in this PR, whose title is "Better typed api". You can definitely open a later PR for it after this one getting merged.

Reply:
Good point, we should stay on track in this PR and decide this later.